Add xml and formatter fix release notes for 2201.9.0 (#8875)
ballerina-platform/ballerina-dev-website: downloads/swan-lake-release-notes/swan-lake-2201.9.0/RELEASE_NOTE.md

### Improvements

#### Allow forward referencing of module-level XML namespace declarations.

> The language supports referencing of module-level XML namespaces which are defined after the usage.

Author: Yes, should it be under the `Backward-incompatible changes`

    ```ballerina
    public function main() {
        xmlns "foo";
        xmlns "bar" as k;
        xml x1 = xml `<item><k:child>C</k:child><child2>D</child2></item>`;
        xml x2 = x1/<*>; // used to return `<child2 xmlns="foo">D</child2>`, now returns `<k:child xmlns="foo" xmlns:k="bar">C</k:child><child2 xmlns="foo">D</child2>`
    }
    ```

- A bug which resulted in XML namespace URI being empty when a constant is used for XML namespace URI in the XML namespace definition has been fixed.

    ```ballerina
    const URI = "http://ballerina.com/";
    xmlns URI as ns0;

    public function main() {
        string s = ns0:foo; // used to result in "{}foo" which now results in {http://ballerina.com/}foo
    }
    ```

- A bug which resulted in a compiler crash or a null value for a variable of a binding pattern which is used within closure has been fixed.

    ```ballerina
    function f1() {
        string[] values = ["a", "b", "c"];
        map<int> data = {
            a: 1
        };
        foreach var [k, _] in data.entries() {
            boolean b = <any> k is (); // which resulted true since k is used in a closure, now returns false. k is now not nil.

>             _ = values.filter(item => item == k);

Reviewer: Can we try to make this a more real-world example?

Make the array an anydata array, use the value instead in the closure, then directly do a nil check and say that the check now evaluates to false unless the value is actually nil.

        }
    }
    ```

> - A bug which resulted in empty XML results in XML navigation when the navigation name pattern contains escape characters has been fixed.

Reviewer: Let's consistently use "empty sequence".

    ```ballerina
    public function main() {
        xml d = xml `
        <person>
            <name>John</name>
            <home-address>some address</home-address>
        </person>`;

        xml x1 = d/<home\-address>; // Used to produce an empty output. Now produce <home-address>some address</home-address>
    }
    ```

> - A bug which resulted in no compilation error on missing required fields in the select clause has been fixed.

Reviewer: ```suggestion
- A bug which resulted in no compilation errors being logged for missing required fields in the `select` clause has been fixed.
```

    ```ballerina
    type Foo record {|
        int i;
        string x;
    |};

    type Bar record {|
        int j;
        string y;
    |};

    type Baz record {|
        Foo[] fooArr;
    |};

    function fn(Baz baz) {

>         int|Bar[] _ = from var foo in baz.fooArr

Reviewer: Avoid using underscore in user-facing content.

Reviewer: Try to make these closer to real-world examples.

>             select { // used to give no compilation error. Now gives a compilation error. missing non-defaultable required record field 'y'

Reviewer: ```suggestion
            select { // Compile-time error now, since the `y` field is not specified.
```

                j: foo.i

>                 // y: foo.x

Reviewer: RmeRmeoveove.

            };
    }
    ```

> - A bug which resulted in swapping the top most comments above imports while formatting is fixed. This fix results in preserving newlines within comment blocks above imports.

Reviewer: ```suggestion
- A bug which resulted in the top-most comments above imports being moved when formatting imports has been fixed. This fix also preserves new lines within comment blocks above imports.
```

    - Before formatting

>     ```ballerina

Reviewer: Can indent, right?

Author: It doesn't seem to be possible to indent the one single level list in an .md file. (when indented it is considered as a code.)

    // Copyright (c) 2024 WSO2 LLC. (http://www.wso2.com).
    //
    // WSO2 LLC. licenses this file to you under the Apache License,
    // Version 2.0 (the "License"); you may not use this file except
    // in compliance with the License.
    // You may obtain a copy of the License at
    //
    // http://www.apache.org/licenses/LICENSE-2.0
    //
    // Unless required by applicable law or agreed to in writing,
    // software distributed under the License is distributed on an
    // "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    // KIND, either express or implied.  See the License for the
    // specific language governing permissions and limitations
    // under the License.

> 

Reviewer: Rethink all the examples and try to make them resemble real-world scenarios as much as possible. Also make them as concise as possible, while preserving readability.

### Bug fixes

## Backward-incompatible changes

> - A bug which resulted in the addition of a default namespace to an XML navigation name pattern, even when the default namespace is defined after it, has been fixed.

Reviewer: ```suggestion
- A bug which resulted in the addition of a default namespace to an XML navigation name pattern, even when the default namespace is declared after it, has been fixed.
```

        xmlns "http://example.com/";
    }
    ```

- A bug which resulted in XML navigation failing to return elements with namespaces different from the default one when using `<*>` to access XML element children has been fixed.

    ```ballerina
    public function main() {
        xmlns "http://example.com/";
        xmlns "https://ballerina.io/" as ns0;

        xml x1 = xml `
        <item>
            <ns0:name>ball</ns0:name>
            <type>t</type>
        </item>`;

        // Previously evaluated to `
        // <type xmlns="http://example.com/">t</type>`,
        // now evaluates to, 
        // `<ns0:name xmlns="http://example.com/" xmlns:ns0="https://ballerina.io">ball</ns0:name>
        // <type xmlns="http://example.com/">t</type>`
        xml x2 = x1/<*>;
    }
    ```

- A bug which resulted in the XML namespace URI being empty when a constant is used in the XML namespace declaration has been fixed.

    ```ballerina
    const URI = "http://ballerina.com/";
    xmlns URI as ns0;

    public function main() {
        // Previously evaluated to "{}attr", now evaluates to {http://ballerina.com/}attr
        string s = ns0:attr;
    }
    ```

> - A bug which resulted in a compiler crash or a nil value for a variable of a binding pattern which is used within closures has been fixed.

Reviewer: ```suggestion
- A bug which resulted in a compiler crash or a nil value for a variable of a binding pattern which is used as a captured variable has been fixed.
```

    ```ballerina
    type Doctor record {
        string name;
        string category;    
    };

    function updateDoctorCategories(Doctor doctor, string[] categories) returns error? {
        var {category} = doctor;
        string cat = category;
         // Previously evaluated to `true` since category is used in a closure, now returns `false`.
        boolean b = <any> category is ();
        if !categories.some(existingCategory => existingCategory == category) {
            categories.push(category);
        }
    }
    ```

> - A bug which resulted in empty XML sequence in XML navigation when the navigation name pattern contains escape characters has been fixed.

Reviewer: ```suggestion
- A bug which resulted in XML navigation evaluating to empty XML sequences when the navigation name pattern contains escape characters has been fixed.
```

        xml x1 = d/<home\-address>; // Previously evaluated to an empty XML sequence, now evaluates to `<home-address>some address</home-address>`
    }
    ```

-  A bug which resulted in no compilation errors being logged for missing required fields in the `select` clause has been fixed.

    ```ballerina
    type Student record {|
        int id;
        string name;
        Enrollment[] enrollments;
    |};

    type Course record {|
        int id;
        string name;
    |};

    type Enrollment record {|
        int id;
        string name;
        int year;
        int grade;
    |};

    function getStudentCourses(Student student) returns int|Course[] {
        int|Course[] courses = from var enrollment in student.enrollments
                            select {
                                // Compile-time error now, since the `name` field is not specified
                                id: enrollment.id
                            };
        return courses;
    }
    ```

- A bug which resulted in the top-most comments above imports being moved when formatting imports has been fixed. This fix also preserves new lines within comment blocks above imports.

>     - Before formatting

Reviewer: Recheck bullet point usage.
